File: JS/mergesort.js
//Mergesort Implementation
function merge(array, left, right) {
  var a = 0;
  while (left.length && right.length){
    if (right[0] < left[0]) {
      array[a++] = right.shift();
    } 
    else {
      array[a++] = left.shift();
    }
  }
  while (left.length){
    array[a++] = left.shift();
  }
  while (right.length){
    array[a++] = right.shift();
  }
};

function mSort(array, temp, leng) {
  if (leng == 1){
    return;
  }
  var m = Math.floor(leng / 2),
  left = temp.slice(0, m),
  right = temp.slice(m);
  mSort(left, array.slice(0, m), m);
  mSort(right, array.slice(m), leng - m);
  merge(array,left, right);
  return array
};

function mergeSort(array) {

  var sorted = mSort(array, array.slice(), array.length);

  return sorted;
};

if (process.argv.length != 3) {
  process.exit(0);
}
var fs = require('fs'); // Read values from file
var array = fs.readFileSync(process.argv[2]).toString().split(/\s+/);
for (var i = 0; i < array.length; ++i) {
  array[i] = parseInt(array[i]);
}

var start = new Date().getTime() ;
mergeSort(array); // mergesort

//console.log(array);
for (var i = 0; i < array.length; ++i) {
  console.log(array[i]);
}
var end = new Date().getTime();
console.log(((end-start)));

File: Haskell/mergesort.hs
import Text.Printf
import Control.Monad
import Control.Exception
import Control.Parallel.Strategies
import System.IO
import System.CPUTime
import System.Environment
import Control.DeepSeq

--mergesort haskell implementation
lim :: Int
lim = 10^6

subDivide :: [Int] -> ([Int],[Int])
subDivide (a:b:rest) = (a:left, b:right) where
                       (left,right) = subDivide rest 
subDivide [a] = ([a],[])
subDivide []  = ([],[])

merge :: [Int] -> [Int] -> [Int]
merge left []  = left
merge [] right = right
merge left@(l:ls) right@(r:rs)
    | l <= r    = l : merge ls right
    | otherwise = r : merge left rs

mergeSort :: [Int] -> [Int]
mergeSort []     = []
mergeSort [arr]  = [arr]
mergeSort array  = merge (mergeSort left) (mergeSort right) where
                   (left,right) = subDivide array

f :: [String] -> [Int]
f = map read

main = do
    [n] <- getArgs
    let list = []
    handle <- openFile n ReadMode
    contents <- hGetContents handle
    let singlewords = words contents
<<<<<<< HEAD
        list = f singlewords    

    let sorted = mergeSort list


    printf "Unsorted List:"
    print list
    printf "Sorted List:"
    start <- getCPUTime
    print sorted
    end <- getCPUTime
    let diff = (fromIntegral (end - start)) / (10^12)

    printf "Computation time: %0.9f sec\n" (diff :: Double)
    printf "Individual time: %0.9f sec\n" (diff / fromIntegral lim :: Double)
=======
        list = f singlewords

    start <- getCPUTime
    let sorted = mergeSort list
    end <- sorted `deepseq` getCPUTime
    let diff = (end - start) `div` (10^9)
    printf "%d\n" (diff :: Integer)
    hClose handle
>>>>>>> FETCH_HEAD
